ui/segmented-bar.reel/segmented-bar.js

```diff
@@ -0,0 +1,132 @@
+    },
+    /**
+     * contentController accepts a 'RangeController' or 'TreeController', the
+     * contentController will default to a RangeController. The TreeController
+     * serves to flatten out a tree to be portrayed as a bar.
+     */
+    contentController: {value: null},
+
+    labelKeyName: {value: null},
+    valueKeyName: {value: null},
+    remainingLabelName: {value: null},
+    remainingValue: {value: null},
+
+    _maxValue: {value: null},
+    maxValue: {
+        get: function () {return this._maxValue},
+        set: function (value) {
+            this._maxValue = value;
+        }
+    },
+
+    /**
+     * Orientation of the bar will be horizontal by default.
+     * The orientation is set to automatically adjust according to the value
+     * of the screen width.
+     */
+    orientation: {value: null},
+    /**
+     * Allows user to override orientation.
+     * TODO: needed?
+     */
+    orientationOverride: {value: null},
+    /**
+     * TODO: needed?
+     */
+    orientationBreakpoint: {value: null},
+
+    /**
+     * logic for RangeController
+     * accepts an RangeController to be projected into the SegmentedBar
+     * TODO: needed? or just an array?
+     */
+
+    /**
+     * logic to flatten TreeController
+     * serves to flatten out a tree to be projected into the SegmentedBar
+     */
+    _flattenTreeController: {value: function () {}},
+
+    enterDocument: {value: function () {
+        this._constructRemainingSegment();
+        // TODO: perf review of listening to win event here
+        // TODO: why doesn't this work in prepareForActivationEvents?
+        window.addEventListener("resize", this);
+        this.orientation =
+            window.matchMedia("(max-width: 1000px)").matches ? "vertical": "horizontal";
+    }},
+    exitDocument: {value: function () {
+        // TODO: what if another component needs to listen for this?
+        window.removeEventListener("resize", this);
+    }},
+
+    /**
+     * Calculate remaining value & construct remaining segment
+     * The remaining value is then pushed into this._content array
+     * to be projected as a segment of the SegmentedBar
+     */
+    _constructRemainingSegment: {
+        value: function _constructRemainingSegment() {
+            var sum = 0;
+            for (var i = 0; i < this.content.length; i++) {
+                sum += this.content[i][this.valueKeyName];
+            }
+
+            // safe guard against user passing in values that total to be greater than max
+            if (sum > this.maxValue) {
+                console.warn(
+                    "SegmentedBar content values total " + sum +
+                    ", maxValue overridden from " + this.maxValue +
+                    " to match.");
+                this.maxValue = sum;
+            } else {
+                this.remainingValue = this.maxValue - sum;
+
+                // construct remaining segment
+                var remainingSegment = {};
```

Comment:
The user story text above is confusing... Is this from the point of view of a developer using the SegmentedBar, as suggested by the context, or from the point of someone using an application that has a SegmentedBar, as suggested by the "As a user" text? Also, this looks a bit like the [format of user stories](http://en.wikipedia.org/wiki/User_story#Format), but it's actually almost the opposite (stating what is required of the interested party, not what is desired by them).

Anyway, I'm going to guess that what we're discussing here is whether the SegmentedBar should have a "leftover value" segment specified through a "totalValue" or "total" property. The term "maxValue" would be wrong for this, by the way, because it is actually not the maximum of any value provided by the API caller. It could be considered the maximum of the total of values provided by API callers, but even that's not clear because that depends on what happens when the total of all value exceeds this "total" value... Oh the confusion!

I don't know yet if we have a need for a leftover segment in Contour. In the absence of a known need, and in the interest of simplicity, I would strongly suggest we not support a leftover segment for now in SegmentedBar. It's not clear whether we should ever support it, because: 1) A leftover can alway be represented as an extra element in the data passed to a leftover-less SegmentedBar; and 2) It makes everything so confusing, changing "a bar split into segments" into "a bar that represents an amount, which sub-amounts represented as segments on that bar and leftover value represented as a different type of segment, and somewhat undefined behavior if the total of the sub-amounts exceeds the bar's amount". The later may be a use case of SegmentedBar, but it would be easier for everyone including ourselves to keep the definition of the SegmentedBar as the former :-).

By the way, here's how easy it would be to represent a leftover in data passed in to the simpler version of SegmentedBar... if that's ever needed:
```
segmentedBarData: {
    get: function() {
        return {children: [this.valueTree, {value: this.capacity - this.valueTree.valueTotal}]}
    }
}
```